subsys/net/ip/net_pkt.c

```diff
@@ -1686,9 +1688,21 @@ int net_pkt_get_src_addr(struct net_pkt *pkt, struct sockaddr *addr,
 		proto = NET_IPV6_HDR(pkt)->nexthdr;
 
 		if (IS_ENABLED(CONFIG_NET_TCP) && proto == IPPROTO_TCP) {
-			addr6->sin6_port = net_pkt_tcp_data(pkt)->src_port;
+			struct net_tcp_hdr hdr;
+
+			if (!net_tcp_get_hdr(pkt, &hdr)) {
```

Comment:
The documentation of net_tcp_get_hdr() seems to be misleading as the net_tcp_get_hdr() does not work like this normally. So if the TCP header fits the fragment, then the function does not populate the hdr variable. This means that the TCP header needs to be accessed via the returned pointer.

You need to write the code like this:
```
struct net_tcp_hdr hdr, *tcp_hdr;

tcp_hdr = net_tcp_get_hdr(pkt, &hdr);
if (!tcp_hdr) {
      return -EINVAL;
}

addr6->sin6_port = tcp_hdr->src_port;
```
Same applies to UDP header handling below.

I will send a patch that documents this behavior better for TCP and UDP header functions.